File: Assignment-9/hotel_server.py
from xmlrpc.server import SimpleXMLRPCServer

class HotelBookingServer:
    def __init__(self):
        self.bookings = {}  # guest_name -> room_number
        self.room_counter = 1  # simple incremental room assignment

    def book_room(self, guest_name):
        if guest_name in self.bookings:
            return f"Guest '{guest_name}' already has room {self.bookings[guest_name]} booked."
        room_number = self.room_counter
        self.bookings[guest_name] = room_number
        self.room_counter += 1
        return f"Room {room_number} successfully booked for '{guest_name}'."

    def cancel_booking(self, guest_name):
        if guest_name in self.bookings:
            room_number = self.bookings.pop(guest_name)
            return f"Booking for guest '{guest_name}' (Room {room_number}) cancelled."
        else:
            return f"No booking found for guest '{guest_name}'."

    def view_bookings(self):
        return self.bookings

# Setup Server
server = SimpleXMLRPCServer(("localhost", 9000))
print("Hotel Booking Server is running on port 9000...")

hotel_server = HotelBookingServer()
server.register_instance(hotel_server)

# Keep server running
server.serve_forever()

File: Assignment-9/src/HotelServer.java
import java.rmi.registry.LocateRegistry;
import java.rmi.registry.Registry;
import java.rmi.server.UnicastRemoteObject;
import java.util.HashMap;
import java.util.Map;

public class HotelServer implements HotelBooking {
    private final Map<Integer, String> bookings = new HashMap<>();

    public HotelServer() { }

    @Override
    public synchronized boolean bookRoom(String guestName, int roomNo) {
        if (bookings.containsKey(roomNo)) return false;
        bookings.put(roomNo, guestName);
        System.out.println("Booked room " + roomNo + " for " + guestName);
        return true;
    }

    @Override
    public synchronized boolean cancelBooking(String guestName, int roomNo) {
        if (guestName.equals(bookings.get(roomNo))) {
            bookings.remove(roomNo);
            System.out.println("Cancelled room " + roomNo + " for " + guestName);
            return true;
        }
        return false;
    }

    public static void main(String[] args) {
        try {
            HotelServer obj = new HotelServer();
            HotelBooking stub = (HotelBooking) UnicastRemoteObject.exportObject(obj, 0);
            Registry registry = LocateRegistry.createRegistry(5050);
            registry.rebind("HotelBookingService", stub);
            System.out.println("HotelServer ready on port 5050.");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}